### PR TITLE
Fix segfault when using Units::equivalent/compatible API

### LINKED ISSUE #1143: Units.compatible() and Units.equivalent() crash
`libcellml version 0.4.0`

I somehow redefined a standard unit. When I used Units.compatible() and Units.equivalent() to check such units, the python terminates without an error or a warning. The following Python code can reproduce the issue:

```python
from libcellml import Component, Model, Units, Variable
model_name = 'unit_compatibletest'
model = Model(model_name)
component=Component(model_name)
component_param=Component(model_name+'_param')
model.addComponent(component)
model.addComponent(component_param)
const_name = 'T'
var_const=Variable(const_name)
unit_name = 'kelvin'
u=Units(unit_name)
u.addUnit(Units.StandardUnit.KELVIN)
model.addUnits(u)
var_const.setUnits(u)
param_const = var_const.clone()
param_const.setInitialValue(293)
model.component(component.name()).addVariable(var_const)
model.component(component_param.name()).addVariable(param_const)
print(f'1 {var_const.name()} units: {param_const.units().name()}')
Units.compatible(var_const.units(), param_const.units()) # where the python terminates without an error or a warning
Units.equivalent(var_const.units(), param_const.units()) # where the python terminates without an error or a warning
print(f'2 {var_const.name()} units: {param_const.units().name()}')
```

### PR DESCRIPTION
**Pull request by hsorby**
Fixes #1143.

### CODE REVIEW

File: src/units.cpp [hunk 1]
```diff
@@ -632,7 +632,7 @@ bool Units::requiresImports() const
         for (size_t u = 0; u < unitCount(); ++u) {
             unitAttributes(u, ref, prefix, exponent, multiplier, id);
             auto child = model->units(ref);
-            if (child == nullptr) {
+            if (child == nullptr || this == child.get()) {
```
Comment (agarny) [inline src/units.cpp]: Parentheses: `if ((child == nullptr) || (this == child.get())) { `.

File: tests/units/units.cpp [hunk 2]
```diff
@@ -2760,3 +2760,32 @@ TEST(Units, validateConnectionsWithBaseUnits)
...
+    libcellml::UnitsPtr units = libcellml::Units::create("kelvin");
+    units->addUnit(libcellml::Units::StandardUnit::KELVIN);
+
+    libcellml::UnitsPtr unitsClone = units->clone();
+
+    model->addUnits(units);
+    variable->setUnits(units);
+
+    libcellml::VariablePtr variableParam = variable->clone();
+    variableParam->setInitialValue(293);
```
Comment (agarny) [inline tests/units/units.cpp]: Not needed for this test.

File: tests/units/units.cpp [hunk 3]
```diff
@@ -2760,3 +2760,32 @@ TEST(Units, validateConnectionsWithBaseUnits)
...
+
+TEST(Units, equivalentUnitsMatchingBuiltinUnitsFromVariable)
+{
+    libcellml::ModelPtr model = libcellml::Model::create("unit_compatible_test");
+
+    libcellml::VariablePtr variable = libcellml::Variable::create("T");
+    libcellml::UnitsPtr units = libcellml::Units::create("kelvin");
+    units->addUnit(libcellml::Units::StandardUnit::KELVIN);
+
+    libcellml::UnitsPtr unitsClone = units->clone();
```
Comment (agarny) [inline tests/units/units.cpp]: Not needed for this test.

**Review by agarny — CHANGES_REQUESTED** [review]
Looks good. Just a few minor changes.

File: tests/units/units.cpp (hunk 2 above)
Comment (hsorby) [reply tests/units/units.cpp]: [45e8f05](https://github.com/cellml/libcellml/pull/1145/commits/45e8f056efd9d4a6f4898d87fa005cf50caa3bf6)

File: src/units.cpp (hunk 1 above)
Comment (hsorby) [reply src/units.cpp]: [2b8c06d](https://github.com/cellml/libcellml/pull/1145/commits/2b8c06d41a5400303f6171ed91d5d69ff035bcba)

File: tests/units/units.cpp (hunk 3 above)
Comment (hsorby) [reply tests/units/units.cpp]: [45e8f05](https://github.com/cellml/libcellml/pull/1145/commits/45e8f056efd9d4a6f4898d87fa005cf50caa3bf6)

**Review by agarny — APPROVED** [review]

**Review by nickerso — APPROVED** [review]

**Pull request merged**